ui/components/FluxRuntime.tsx

```diff
@@ -5,27 +5,104 @@ import Flex from "../components/Flex";
 import { Crd, Deployment } from "../lib/api/core/types.pb";
 import ControllersTable from "./ControllersTable";
 import CrdsTable from "./CrdsTable";
+import FluxVersionsTable, { FluxVersion } from "./FluxVersionsTable";
+import { routeTab } from "./KustomizationDetail";
 import SubRouterTabs, { RouterTab } from "./SubRouterTabs";
+import Text from "./Text";
+
+const FluxVersionText = styled(Text)`
+  font-weight: 700;
+  margin-bottom: ${(props) => props.theme.spacing.medium};
+
+  span {
+    color: ${(props) => props.theme.colors.neutral40};
+    font-weight: 400;
+    margin-left: ${(props) => props.theme.spacing.xs};
+  }
+`;
 
 type Props = {
   className?: string;
   deployments?: Deployment[];
   crds?: Crd[];
+  supportMultipleFlux?: boolean;
 };
+const fluxVersionLabel = "app.kubernetes.io/version";
 
-function FluxRuntime({ className, deployments, crds }: Props) {
+function FluxRuntime({
+  className,
+  deployments,
+  crds,
+  supportMultipleFlux = false,
+}: Props) {
   const { path } = useRouteMatch();
+  const tabs: Array<routeTab> = [
+    {
+      name: "Controllers",
+      path: `${path}/controllers`,
+      component: () => {
+        return <ControllersTable controllers={deployments} />;
+      },
+      visible: true,
+    },
+    {
+      name: "CRDs",
+      path: `${path}/crds`,
+      component: () => {
+        return <CrdsTable crds={crds} />;
+      },
+      visible: true,
+    },
+  ];
 
+  if (supportMultipleFlux) {
+    const fluxVersions: Array<FluxVersion> = [];
+    deployments.forEach((d) => {
```

Comment:
This `.forEach` is mutating an array that it also refers to, which is a bit circular and confusing IMO.

Instead, what about:

```
    const allDeploymentVersions = _.map(deployments, (d) => ({
      version: d.labels[fluxVersionLabel],
      clusterName: d.clusterName,
      namespace: d.namespace,
    }));

    const uniq = _.uniqBy(allDeploymentVersions, "clusterName");
```
I think this would indicate the intent a little better. The `lodash` funcs can be replaced with their native (worse :wink:) equivalents.

Reply:
`.forEach` would be a problem if we make a change to the element in the `callbackFn` but it only apply once for each array element and we don't change anything to the deployments array.

`lodash` is amazing :joy: i know, but i tried to test how worse `ES6` is to `lodash` by 

```import _ from 'lodash'
const fluxVersionLabel = "app.kubernetes.io/version";
const dd = [
  {
    "name": "helm-controller",
    "namespace": "flux-system",
    "clusterName": "management",
    "labels": {
      "app.kubernetes.io/version": "v0.30.2",
    }
  },
  {
    "name": "kustomize-controller",
    "namespace": "flux-system",
    "clusterName": "management",
    "labels": {
      "app.kubernetes.io/version": "v0.30.2",
    }
  },
  {
    "name": "notification-controller",
    "namespace": "flux-system",
    "clusterName": "management",
    "labels": {
      "app.kubernetes.io/version": "v0.30.2",
    }
  },
  {
    "name": "source-controller",
    "namespace": "flux-system",
    "clusterName": "management",
    "labels": {
      "app.kubernetes.io/version": "v0.30.2",
    }
  },
  {
    "name": "helm-controller",
    "namespace": "flux-system",
    "clusterName": "default/tw-cluster-2",
    "labels": {
      "app.kubernetes.io/version": "v0.32.0",
    }
  },
  {
    "name": "kustomize-controller",
    "namespace": "flux-system",
    "clusterName": "default/tw-cluster-2",
    "labels": {
      "app.kubernetes.io/version": "v0.32.0",
    }
  },
  {
    "name": "notification-controller",
    "namespace": "flux-system",
    "clusterName": "default/tw-cluster-2",
    "labels": {
      "app.kubernetes.io/version": "v0.32.0",
    }
  },
  {
    "name": "source-controller",
    "namespace": "flux-system",
    "clusterName": "default/tw-cluster-2",
    "labels": {
      "app.kubernetes.io/version": "v0.32.0",
    }
  }
]
const deployments=[...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd,...dd]
function lodashFn() {
  const startTime = performance.now()
  const allDeploymentVersions = _.map(deployments, (d) => ({
    version: d.labels[fluxVersionLabel],
    clusterName: d.clusterName,
    namespace: d.namespace,
  }));
  const uniq = _.uniqBy(allDeploymentVersions, dd=>[dd.version,dd.clusterName].join());
  const endTime = performance.now()
  console.log(`lodashFn: ${endTime - startTime} ms`)  
}

function es6For() {
  const startTime = performance.now()
  const fluxVersions: any = {};
  for (let index = 0; index < deployments.length; index++) {
    const d = deployments[index];
    const fv = d.labels[fluxVersionLabel];
    const k = `${fv}${d.clusterName}`;
    if (!fluxVersions[k]) {
      fluxVersions[k] = {
        version: fv,
        clusterName: d.clusterName,
        namespace: d.namespace,
      };
    }
  }
  const endTime = performance.now()
  console.log(`es6For: ${endTime - startTime} ms`)  
}
function es6Foreach(){
  const startTime = performance.now()
  const fluxVersions: any = {};
  deployments.forEach((d) => {
    const fv = d.labels[fluxVersionLabel];
    const k = `${fv}${d.clusterName}`;
    if (!fluxVersions[k]) {
      fluxVersions[k] = {
        version: fv,
        clusterName: d.clusterName,
        namespace: d.namespace,
      };
    }
  });
  const endTime = performance.now()
  console.log(`es6Foreach: ${endTime - startTime} ms`)  
} 

``` 

Calling interactively each `10 sec` and results comes as :

```
es6For: 0.599999999627471 ms
es6Foreach: 0.599999999627471 ms
lodashFn: 1.300000000745058 ms

es6For: 0.40000000037252903 ms
es6Foreach: 0.599999999627471 ms
lodashFn: 0.6000000014901161 ms

es6For: 0.599999999627471 ms
es6Foreach: 0.5 ms
lodashFn: 1.199999999254942 ms

es6For: 0.5 ms
es6Foreach: 0.2999999988824129 ms
lodashFn: 1.5 ms

es6For: 0.5 ms
es6Foreach: 0.7000000011175871 ms
lodashFn: 1.300000000745058 ms

es6For: 1.099999999627471 ms
es6Foreach: 1.699999999254942 ms
lodashFn: 2.799999998882413 ms

es6For: 0.6000000014901161 ms
es6Foreach: 0.8000000007450581 ms
lodashFn: 1.099999999627471 ms

es6For: 2.800000000745058 ms
es6Foreach: 1.6000000014901161 ms
lodashFn: 2.5 ms

es6For: 1.900000000372529 ms
es6Foreach: 1.8999999985098839 ms
lodashFn: 8.199999999254942 ms

es6For: 0.19999999925494194 ms
es6Foreach: 0.599999999627471 ms
lodashFn: 1.400000000372529 ms
```

#### decrease depolyments :
```
const deployments=[...dd,...dd,...dd,...dd,...dd,...dd]

## Results
es6For: 0.40000000037252903 ms
es6Foreach: 0.5 ms
lodashFn: 0.8999999985098839 ms

es6For: 0.09999999962747097 ms
es6Foreach: 0.09999999962747097 ms
lodashFn: 0.2999999988824129 ms

es6For: 0.09999999962747097 ms
es6Foreach: 0 ms
lodashFn: 0.09999999962747097 ms

es6For: 0.09999999962747097 ms
es6Foreach: 0 ms
lodashFn: 0.09999999962747097 ms

es6For: 0.09999999962747097 ms
es6Foreach: 0.19999999925494194 ms
lodashFn: 0.19999999925494194 ms

es6For: 0 ms
es6Foreach: 0.09999999962747097 ms
lodashFn: 0 ms

es6For: 0.2000000011175871 ms
es6Foreach: 0.09999999962747097 ms
lodashFn: 0.19999999925494194 ms

es6For: 0 ms
es6Foreach: 0 ms
lodashFn: 0.09999999962747097 ms

es6For: 0.09999999962747097 ms
es6Foreach: 0 ms
lodashFn: 0.10000000149011612 ms

es6For: 0 ms
es6Foreach: 0.09999999962747097 ms
lodashFn: 0 ms

```

Reply:
Premature optimization. Code readability and not exploding on `undefined` are the advantages of lodash. In any case, the same readability can be achieved without `lodash` if that is truly a sticking point.